### PR TITLE
Partial Retract WOH

### PR DESCRIPTION
**Pull request by gregorydlogan**
Adding partial retract WOH, which enables retraction of specific flavour(s) and tags rather than pulling down the whole mediapackage.

Work sponsored by University of Cape Town

### CODE REVIEW

**Comment by gregorydlogan** [comment]
This is filed against 8.x at https://github.com/cilt-uct/opencast/pull/57

File: docs/guides/admin/docs/workflowoperationhandlers/retract-partial-woh.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+|retract-flavors            |Which flavor(s) to retract.  Use a comma to separate multiple flavors.                       |
+|retract-tags               |Which tags(s) to retract.  Use a comma to separate multiple tags.                            |
+
+
+Operation Example
+-----------------
+
+```xml
+<operation
+    id="retract-engage"
```
Comment (NgoniChoga) [inline docs/guides/admin/docs/workflowoperationhandlers/retract-partial-woh.md]: Hi @gregorydlogan, a little "gotcha" here. Should be retract-partial.

Comment (gregorydlogan) [reply docs/guides/admin/docs/workflowoperationhandlers/retract-partial-woh.md]: Ah, good catch!

File: modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PartialRetractEngageWorkflowOperationHandler.java [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+  public WorkflowOperationResult start(WorkflowInstance workflowInstance, JobContext context)
+          throws WorkflowOperationException {
+    WorkflowOperationInstance op = workflowInstance.getCurrentOperation();
+
+    // Check which tags have been configured
+    String retractTargetTags = StringUtils.trimToEmpty(op.getConfiguration(RETRACT_TAGS));
+    String retractTargetFlavors = StringUtils.trimToNull(op.getConfiguration(RETRACT_FLAVORS));
+
+    String[] retractTags = StringUtils.split(retractTargetTags, ",");
+    String[] retractFlavors = StringUtils.split(retractTargetFlavors, ",");
```
Comment (CGreweling) [inline modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PartialRetractEngageWorkflowOperationHandler.java:78]: is this null safe?
beacause of trimtonull

Comment (gregorydlogan) [reply modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PartialRetractEngageWorkflowOperationHandler.java:78]: Derp, that's what I get for copy and paste :)

**Review by lkiesow — CHANGES_REQUESTED** [review]
Generally looks okay, but when I start Opencast, I get:

```
2020-09-30 17:49:47,047 | ERROR | (R6LogServiceLogger:50) - bundle opencast-distribution-workflowoperation:9.0.0.SNAPSHOT (87)[org.opencastproject.workflow.handler.distribution.PartialRetractEngageWorkflowOperationHandler(45)] : bind method [setStreamingDistributionService] not found; Component will fail
2020-09-30 17:49:47,048 | ERROR | (R6LogServiceLogger:50) - bundle opencast-distribution-workflowoperation:9.0.0.SNAPSHOT (87)[org.opencastproject.workflow.handler.distribution.PartialRetractEngageWorkflowOperationHandler(45)] : bind method [setStreamingDistributionService] not found
```

Also, maybe use OSGi annotations if that is possible instead of the XML definition? That problem could probably have been avoided by that

**Comment by gregorydlogan** [comment]
@lkiesow should be fixed.  Don't have time to get at the annotations today unfortunately.

File: docs/guides/admin/docs/workflowoperationhandlers/retract-partial-woh.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+```xml
+<operation
+    id="retract-partial"
+    fail-on-error="true"
+    exception-handler-workflow="partial-error"
+    description="Retracting elements flavored with presentation and tagged with preview from Engage">
+  <configurations>
+    <configuration key="retract-flavors">presentation/*</configuration>
+    <configuration key="retract-tags">preview</configuration>
+  <configurations>
```
Comment (lkiesow) [inline docs/guides/admin/docs/workflowoperationhandlers/retract-partial-woh.md:32]:
```suggestion
  </configurations>
```

**Review by lkiesow — CHANGES_REQUESTED** [review]
Does what it's supposed to do.

There are only two documentation issues.
One is the invalid XML in the example below (see suggestion), the other one is that the new operation is missing from the list of operation handlers:
https://github.com/opencast/opencast/blob/develop/docs/guides/admin/docs/workflowoperationhandlers/index.md

Feel free to merge this yourself after fixing that. I don't thing another review is necessary for it.

**Pull request merged**